xdsl/backend/riscv/lowering/convert_memref_to_riscv.py

```diff
@@ -296,6 +367,100 @@ def match_and_rewrite(self, op: memref.GetGlobal, rewriter: PatternRewriter):
         )
 
 
+class ConvertMemrefSubviewOp(RewritePattern):
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: memref.Subview, rewriter: PatternRewriter):
+        # Assumes that the operation is valid, meaning that the subview is indeed a
+        # subview, and that if the offset is stated in the layout attribute, then it's
+        # correct.
+
+        source = op.source
+        result = op.result
+        source_type = source.type
+        assert isinstance(source_type, MemRefType)
+        source_type = cast(MemRefType[Attribute], source_type)
+        result_type = cast(MemRefType[Attribute], result.type)
+
+        result_layout_attr = result_type.layout
+        if isinstance(result_layout_attr, NoneAttr):
+            # When a subview has no layout attr, the result is a perfect subview at offset
+            # 0.
+            rewriter.replace_matched_op(
+                UnrealizedConversionCastOp.get((source,), (result_type,))
+            )
+            return
+
+        if not isinstance(result_layout_attr, StridedLayoutAttr):
+            raise DiagnosticException("Only strided layout attrs implemented")
+
+        # We still need to check that the layout attribute strides are contiguous in
+        # memory.
+        result_shape = result_type.get_shape()
+        result_strides = tuple(result_layout_attr.get_strides())
+
+        identity_strides = tuple(
+            accumulate(reversed(result_shape), lambda acc, new: acc * new, initial=1)
+        )[::-1]
+
+        if identity_strides[-len(result_strides) :] != result_strides:
+            # The subview may be rank reducing, we want the suffix of the original strides
+            raise DiagnosticException(
+                "Cannot lower strides that do not result in a contiguous subview"
+            )
```

Comment:
Why is this a limitation? 
And if this is due to our stack not supporting it, shouldn't the issue be caught with appropiate errors in the access operations rather than here?
I'd have expected the `subview` to perform an offset on the bare pointer regardless of the result type.

Reply:
I didn't think too much about it, but you're right, that should be checked in the store/load lowering, I'll add it

- [ ] check layout on store and load lowering

Reply:
I'd also do this in a separate PR given its unrelated to the subview lowering

Reply:
Thinking about it a bit more, the strides are dangerous in that our riscv pointer representation assumes that everything is contiguous. I think it's still a good idea to check for it when lowering subview operations, just to increase the chances of catching the errors early.

Reply:
Do you know what operations support strides and which don't?

I am very much passing non-contigious memref to xdsl (they are basically required if you want to any kind of tiling besides workgroup tiling) and currently investigating as to why the xDSL workflow returns different results than LLVM in Qudditch at HEAD. Do note that the IR I am passing does not contain any subviews at all, just linalg operations, so these checks would not help in that case (which is why I'd rather see them at the memref consumer that cannot support them, rather than the memref producer that can)

Reply:
OK then let's work out how best to deal with non-contiguous memrefs, it would be good to have a better idea of the kinds of patterns we need to support.